mozilla-services/consvc-shepherd
# [AE-486] First-pass React development setup for AdOps Dashboard

## References

[AE-486](https://mozilla-hub.atlassian.net/browse/AE-486)

## Problem Statement

We need to set up a new React frontend in shepherd for the AdOps Dashboard, as well as a separate place for our API endpoints that will drive this UI. We want to be able to develop it alongside the existing shepherd templated frontend and admin functionality.

## Proposed Changes

This change adds a first-pass development setup for a React frontend and json REST API. 

Several decisions about tooling and repo organization are made in this PR. I'll call them out explicitly and I welcome any discussion/suggestions around these decisions.

- New directory for the new frontend, called `./dashboard`
- In `./dashboard`, Set up Vite + React + Typescript with SWC (speedier, newer than Babel)
- Set up `Dockerfile` and `docker-compose` with the new dashboard app so a single `docker-compose` command from the project root spins up everything. Changes developers make are hot-reloaded into the docker container
- Sets up a json REST API on the django end to serve data to the dashboard. The dashboard overviews aren't really RESTful resources, but I wanted to go with a very standard API choice with lots of tooling/support.
- Provisional App.jsx that demos that we can fetch from the API
- `react-query` to wrap up and consolidate common data-fetching concerns like cacheing, retries. This can also give some state management but I haven't gotten there yet.

## Verification Steps

Run `docker-compose up --build` from the root of the repo
Visit http://0.0.0.0:7001/api/v1/products to see Django's API debug pages
Visit http://0.0.0.0:5173/ and make sure you can see the React UI render a list of the BoostrProducts

## Next Steps

This is very bare bones currently so it can unlock others developing on the dashboard, and it still needs quite a bit of additional setup, which I'll add incrementally in subsequent PRs.

- jest test setup and first-pass unit test
- Add dashboard linting and tests to the existing Makefile
- Build for production and serve React app from Django

## Visuals

<img width="889" alt="Screenshot 2024-09-11 at 4 07 20 PM" src="https://github.com/user-attachments/assets/ba2292d3-b6f0-4f7e-bfb7-e4493cfed7ff">
<img width="1470" alt="Screenshot 2024-09-11 at 4 09 44 PM" src="https://github.com/user-attachments/assets/fdf4c912-edc2-4584-81bb-d1076178ad8e">
